src/Stratis.Bitcoin/Consensus/Validators/BlockValidator.cs

```diff
@@ -37,39 +48,117 @@ public interface IHeaderValidator
-            throw new System.NotImplementedException();
+            this.logger.LogTrace("({0}:'{1}')", nameof(chainedHeader), chainedHeader);
+
+            var validationContext = new ValidationContext { Block = block };
+
+            this.consensusRules.IntegrityValidationAsync(validationContext, null);
+
+            this.logger.LogTrace("(-)");
         }
     }
 
     // <inheritdoc />
     public class PartialValidation : IPartialValidation
     {
+        private readonly IConsensusRules consensusRules;
+        private readonly AsyncQueue<PartialValidationItem> asyncQueue;
+        private readonly ILogger logger;
+
+        public PartialValidation(IConsensusRules consensusRules, ILoggerFactory loggerFactory)
+        {
+            this.consensusRules = consensusRules;
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+
+            this.asyncQueue = new AsyncQueue<PartialValidationItem>(this.OnEnqueueAsync);
+        }
+
+        private async Task OnEnqueueAsync(PartialValidationItem item, CancellationToken cancellationtoken)
+        {
+            this.logger.LogTrace("({0}:'{1}')", nameof(item), item);
+
+            var validationContext = new ValidationContext {Block = item.ChainedHeaderBlock.Block};
+
+            await this.consensusRules.PartialValidationAsync(new ValidationContext { Block = item.ChainedHeaderBlock.Block, ChainedHeader = item.ChainedHeaderBlock.ChainedHeader }, null);
+
+            var partialValidationResult = new PartialValidationResult
+            {
+                ChainedHeaderBlock = item.ChainedHeaderBlock,
+                BanDurationSeconds = validationContext.BanDurationSeconds,
+                BanReason = validationContext.Error != null ? $"Invalid block received: {validationContext.Error.Message}" : string.Empty,
+                Succeeded = validationContext.Error != null
+            };
+
+            await item.PartialValidationCompletedAsyncCallback(partialValidationResult).ConfigureAwait(false);
+
+            this.logger.LogTrace("(-)");
+        }
+
         // <inheritdoc />
         public void StartPartialValidation(ChainedHeaderBlock chainedHeaderBlock, OnPartialValidationCompletedAsyncCallback onPartialValidationCompletedAsyncCallback)
         {
-            throw new System.NotImplementedException();
+            this.logger.LogTrace("({0}:'{1}')", nameof(chainedHeaderBlock), chainedHeaderBlock);
+
+            this.asyncQueue.Enqueue(new PartialValidationItem()
+            {
+                ChainedHeaderBlock = chainedHeaderBlock,
+                PartialValidationCompletedAsyncCallback = onPartialValidationCompletedAsyncCallback
+            });
+
+            this.logger.LogTrace("(-)");
+        }
+
+        public class PartialValidationItem
+        {
+            public ChainedHeaderBlock ChainedHeaderBlock { get; set; }
+
+            public OnPartialValidationCompletedAsyncCallback PartialValidationCompletedAsyncCallback { get; set; }
```

Comment:
Maybe instead of storing a callback for each block we should initialize an instance of the validator with the callback? 

We don't expect more than 1 component to call partial validation using the same instance of the partial validator (also most likely only CM will call it, not sure if other components need it).

Reply:
Same as the puller? perhaps yes, next PR but lets discuss.